### DISCUSSION TITLE
Welcome to IrpTransmogrifier Discussions!

### DISCUSSION THREAD

**Discussion by bengtmartensson** [bengtmartensson/IrpTransmogrifier#194] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by johno3421** [comment]
Hello!

I'm currently developing an IR Remote Control App for my LG G4 phone, using its internal IR support - my intention is to try and integrate it with the various files that I have already created using the JP1 keymap-master-v9.22.xls tool.
For many years I have used the Kameleon 8060 remote ( using the extender ), and have programmed up all of my devices for this remote.
Unfortunately, they are now showing their age, and don't work very reliably any more.

I have created my App using Xamarin.Forms and Microsoft Visual Studio 2019

I have coded up ( in C# ) an implementation of the Sharp protocol, and successfully transmitted the pattern using the Android ConsumerIrManager.transmit method. 👍 
Access to this method is done via the Android Java Bindings Library Mono.Android

As part of my investigations, I came across the IRP Notation, and thought that I would look into this as a method to implement other protocols.
Then - I found IRPMaster, and then IrpTransmogrifier, and started to investigate if I could use it in my App

I have a number of initial points...

Working at the command line of Windows 10...

'mvn install' failed 3 of the tests

So I tried 'mvn install -Dmaven.test.skip=true'

This got me further, but the pom.xml wants to run the *nix script 'mkchecksums.sh'

So I commented this part out, and then got:
Failed to execute goal org.codehaus.mojo:xml-maven-plugin:1.0.2:transform (default) on project IrpTransmogrifier: Execution default of goal org.codehaus.mojo:xml-maven-plugin:1.0.2:transform failed: URI is not absolute

More generally, before I spotted IrpTransmogrifier, I had some success building an Android Java Bindings Library using IRPMaster ( so that I can call into your stuff from C# ), but I am getting quite a lot of errors when I try and create a Bindings Library using the execute jar that you ship in your IrpTransmogrifier-1.2.8-bin.zip package

This is just one example:
1>C:\Users\John\source\repos\IrpTransmogrifier\IrpTransmogrifier\obj\Debug\generated\src\Org.Harctoolbox.Irp.BareIrStream.cs(10,84,10,125): error CS0535: 'BareIrStream' does not implement interface member 'IIrStreamItem.IsEmpty(NameEngine)'

I would be interested in your thoughts on what I am trying to do - do you have any experience of using IrpTransmogrifier on Android / iOS?

**Comment by bengtmartensson** [comment]
Hi, and thanx for your contribution.

> 'mvn install' failed 3 of the tests

Please give more details, and I will investigate. Make sure you are using the current Git sources. For me, they pass.

> This got me further, but the pom.xml wants to run the *nix script 'mkchecksums.sh'

That is strictly speaking only of interest to the maintainer. I have checked in a new `pom.xml` that only executes that script if `/bin/sh` exists, so you do not have to comment out anything anymore. Thanx for pointing this out.

> Failed to execute goal org.codehaus.mojo:xml-maven-plugin:1.0.2:transform (default) on project IrpTransmogrifier: Execution default of goal org.codehaus.mojo:xml-maven-plugin:1.0.2:transform failed: URI is not absolute

I cannot reproduce as I do not get that error. Please help localize.

> 1>C:\Users\John\source\repos\IrpTransmogrifier\IrpTransmogrifier\obj\Debug\generated\src\Org.Harctoolbox.Irp.BareIrStream.cs(10,84,10,125): error CS0535: 'BareIrStream' does not implement interface member 'IIrStreamItem.IsEmpty(NameEngine)'

This is not my code; there is no .../IrpTransmogrifier/IrpTransmogrier/obj/Debug/... in my build process, and there is also no interface IIrStreamItem in my code. So you are on your own here.

> ... successfully transmitted the pattern using the Android ConsumerIrManager.transmit method. +1
Access to this method is done via the Android Java Bindings Library Mono.Android
>
> I would be interested in your thoughts on what I am trying to do - do you have any experience of using IrpTransmogrifier on Android / iOS?

The current project does not do any hardware access nor does it contain any hardware dependencies. Instead, check out the the project [HarcHardware](https://github.com/bengtmartensson/HarcHardware), which sits on top of IrpTransmogrifer, and provides hardware access, for usage in for example [IrScrutinizer](https://github.com/bengtmartensson/IrScrutinizer). 

I would very much welcome contributions in that direction.

**Comment by johno3421** [comment]
Hi Bengt,

 

Thanks for your response.

 

I have now had the opportunity to look further into your package, and have been experimenting!

Attached is the output from my build, showing the test failures

 

Having started over after removing the mkchecksums.sh reference, and running  ‘mvn install -Dmaven.test.skip=true’, I have no issues building the package.

 

As I am trying to use your package on an Android device, I need to build a ‘Java Bindings Library’, and it is here that I have spent most of my time.

The CS0535 error I reported I now understand (along with many others), and I have been able to generate a bindings library, 

which I can now add as a reference to my Xamarin Android App – all good!

 

However, I have run into a (quite serious – for me) problem, which relates to the XML aspects of IrpTransmogrifier – I don’t have access to the javax.xml package,

so cannot setup my IrpDatabase class from the supplied IrpProtocols.xml file - \org\harctoolbox\xml is a no go area L

 

It is my feeling that the best way forward for me would be to add a JSON option to the list of IrpDatabase constructors

 

I haven’t written anything substantial in Java, having spent my ( now retired from ) career writing in various assemblers, C, C++ and C#, but I can give it a go!

I’m guessing that adding this isn’t something that you would really want to do…

 

Having already established that I can output IR signal using the Android phone hardware, I don’t have any hardware issues,

so I won’t be looking at HarcHardware or IrScrutinizer just yet.

 

However, being locked down at home here in the UK for the foreseeable future, I do have plenty of time

 

Regards,

 

John Oseman

**Comment by bengtmartensson** [comment]
[Please try to format better; use gists for longer enclosures, and no full-quotes. Thank you.]

> Attached is the output from my build, showing the test failures

Created #196, already fixed. Thank you! Try it again.

> I don’t have access to the javax.xml package,

There is really no alternative to fixing this. Possibly you can find another Java ([AdoptOpenJDK](https://github.com/AdoptOpenJDK)?) or you can snarf javax.xml from another Java installation?

> I haven’t written anything substantial in Java, having spent my ( now retired from ) career writing in various assemblers, C, C++ and C#, but I can give it a go!

Java and C# are really not that different.
